File: 1_basic/WindowsForms/Form1.cs
using System;
using System.Drawing;
using System.Windows.Forms;
using Microsoft.Kinect;
using System.Drawing.Imaging;
using System.Runtime.InteropServices;

namespace WindowsForms
{
  public partial class Form1 : Form
  {
    readonly int Bgr32BytesPerPixel = 4;  // ピクセルあたりのバイト数

    public Form1()
    {
      InitializeComponent();

      try {
        // Kinectが接続されているかどうかを確認する
        if ( KinectSensor.KinectSensors.Count == 0 ) {
          throw new Exception( "Kinectを接続してください" );
        }

        // Kinectの動作を開始する
        StartKinect( KinectSensor.KinectSensors[0] );
      }
      catch ( Exception ex ) {
        MessageBox.Show( ex.Message );
        Close();
      }
    }

    /// <summary>
    /// Kinectの動作を開始する
    /// </summary>
    /// <param name="kinect"></param>
    private void StartKinect( KinectSensor kinect )
    {
      kinect.ColorStream.Enable();
      kinect.DepthStream.Enable();
      kinect.SkeletonStream.Enable();

      kinect.AllFramesReady += new EventHandler<AllFramesReadyEventArgs>( kinect_AllFramesReady );

      kinect.Start();

      // defaultモードとnearモードの切り替え
      comboBoxRange.Items.Clear();
      foreach ( var range in Enum.GetValues( typeof( DepthRange ) ) ) {
        comboBoxRange.Items.Add( range.ToString() );
      }

      comboBoxRange.SelectedIndex = 0;
    }

    /// <summary>
    /// Kinectの動作を停止する
    /// </summary>
    /// <param name="kinect"></param>
    private void StopKinect( KinectSensor kinect )
    {
      if ( kinect != null ) {
        if ( kinect.IsRunning ) {
          kinect.AllFramesReady -= kinect_AllFramesReady;

          kinect.Stop();
          kinect.Dispose();
          kinect = null;

          pictureBoxRgb.Image = null;
          pictureBoxDepth.Image = null;
        }
      }
    }

    /// <summary>
    /// RGBカメラ、距離カメラ、骨格のフレーム更新イベント
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    void kinect_AllFramesReady( object sender, AllFramesReadyEventArgs e )
    {
      try {
        // Kinectのインスタンスを取得する
        KinectSensor kinect = sender as KinectSensor;
        if ( kinect == null ) {
          return;
        }

        // RGBカメラのフレームデータを取得する
        using ( ColorImageFrame colorFrame = e.OpenColorImageFrame() ) {
          if ( colorFrame != null ) {
            // RGBカメラのピクセルデータを取得する
            byte[] colorPixel = new byte[colorFrame.PixelDataLength];
            colorFrame.CopyPixelDataTo( colorPixel );

            // ピクセルデータをビットマップに変換する
            Bitmap bitmap = new Bitmap( kinect.ColorStream.FrameWidth, kinect.ColorStream.FrameHeight,
                        System.Drawing.Imaging.PixelFormat.Format32bppRgb );

            Rectangle rect = new Rectangle( 0, 0, bitmap.Width, bitmap.Height );
            BitmapData data = bitmap.LockBits( rect, ImageLockMode.WriteOnly,
                                    PixelFormat.Format32bppRgb );
            Marshal.Copy( colorPixel, 0, data.Scan0, colorPixel.Length );
            bitmap.UnlockBits( data );

            pictureBoxRgb.Image = bitmap;
          }
        }

        // 書き込み用のビットマップデータを作成(32bit bitmap)
        // 16bpp グレースケールは表示できない
        // 距離カメラのフレームデータを取得する
        using ( DepthImageFrame depthFrame = e.OpenDepthImageFrame() ) {
          if ( depthFrame != null ) {
            // 距離データを画像化して表示
            Bitmap bitmap = new Bitmap( kinect.DepthStream.FrameWidth, kinect.DepthStream.FrameHeight,
                        System.Drawing.Imaging.PixelFormat.Format32bppRgb );

            Rectangle rect = new Rectangle( 0, 0, bitmap.Width, bitmap.Height );
            BitmapData data = bitmap.LockBits( rect, ImageLockMode.WriteOnly,
                                    System.Drawing.Imaging.PixelFormat.Format32bppRgb );
            byte[] gray = ConvertDepthColor( kinect, depthFrame );
            Marshal.Copy( gray, 0, data.Scan0, gray.Length );
            bitmap.UnlockBits( data );

            pictureBoxDepth.Image = bitmap;
          }
        }

        // スケルトンのフレームを取得する
        using ( SkeletonFrame skeletonFrame = e.OpenSkeletonFrame() ) {
          if ( skeletonFrame != null ) {
            Graphics g = Graphics.FromImage( pictureBoxRgb.Image );

            // スケルトンのデータを取得する
            Skeleton[] skeletons = new Skeleton[skeletonFrame.SkeletonArrayLength];
            skeletonFrame.CopySkeletonDataTo( skeletons );

            // トラッキングされているスケルトンのジョイントを描画する
            foreach ( var skeleton in skeletons ) {
              // スケルトンがトラッキング状態(デフォルトモード)の場合は、ジョイントを描画する
              if ( skeleton.TrackingState == SkeletonTrackingState.Tracked ) {
                // ジョイントを描画する
                foreach ( Joint joint in skeleton.Joints ) {
                  // ジョイントがトラッキングされていなければ次へ
                  if ( joint.TrackingState != JointTrackingState.Tracked ) {
                    continue;
                  }

                  // スケルトンの座標を、RGBカメラの座標に変換して円を書く
                  DrawEllipse( kinect, g, joint.Position );
                }
              }
              // スケルトンが位置追跡(ニアモードの)の場合は、スケルトン位置(Center hip)を描画する
              else if ( skeleton.TrackingState == SkeletonTrackingState.PositionOnly ) {
                // スケルトンの座標を、RGBカメラの座標に変換して円を書く
                DrawEllipse( kinect, g, skeleton.Position );
              }
            }
          }
        }
      }
      catch ( Exception ex ) {
        MessageBox.Show( ex.Message );
      }
    }

    private static void DrawEllipse( KinectSensor kinect, Graphics g, SkeletonPoint position )
    {
      const int R = 5;

      ColorImagePoint point = kinect.MapSkeletonPointToColor( position,
        kinect.ColorStream.Format );
      g.DrawEllipse( new Pen( Brushes.Red, R ),
        new Rectangle( point.X - R, point.Y - R, R * 2, R * 2 ) );
    }


    /// <summary>
    /// 距離データをカラー画像に変換する
    /// </summary>
    /// <param name="depthFrame"></param>
    /// <returns></returns>
    private byte[] ConvertDepthColor( KinectSensor kinect, DepthImageFrame depthFrame )
    {
      ColorImageStream colorStream = kinect.ColorStream;
      DepthImageStream depthStream = kinect.DepthStream;

      // 距離カメラのピクセルごとのデータを取得する
      short[] depthPixel = new short[depthFrame.PixelDataLength];
      depthFrame.CopyPixelDataTo( depthPixel );

      // 距離カメラの座標に対応するRGBカメラの座標を取得する(座標合わせ)
      ColorImagePoint[] colorPoint = new ColorImagePoint[depthFrame.PixelDataLength];
      kinect.MapDepthFrameToColorFrame( depthStream.Format, depthPixel,
        colorStream.Format, colorPoint );

      byte[] depthColor = new byte[depthFrame.PixelDataLength * Bgr32BytesPerPixel];
      for ( int index = 0; index < depthPixel.Length; index++ ) {
        // 距離カメラのデータから、プレイヤーIDと距離を取得する
        int player = depthPixel[index] & DepthImageFrame.PlayerIndexBitmask;
        int distance = depthPixel[index] >> DepthImageFrame.PlayerIndexBitmaskWidth;

        // 変換した結果が、フレームサイズを超えることがあるため、小さいほうを使う
        int x = Math.Min( colorPoint[index].X, colorStream.FrameWidth - 1 );
        int y = Math.Min( colorPoint[index].Y, colorStream.FrameHeight - 1 );
        int colorIndex = ((y * depthFrame.Width) + x) * Bgr32BytesPerPixel;

        // プレイヤーがいるピクセルの場合
        if ( player != 0 ) {
          depthColor[colorIndex] = 255;
          depthColor[colorIndex + 1] = 255;
          depthColor[colorIndex + 2] = 255;
        }
        // プレイヤーではないピクセルの場合
        else {
          // サポート外 0-40cm
          if ( distance == depthStream.UnknownDepth ) {
            depthColor[colorIndex] = 0;
            depthColor[colorIndex + 1] = 0;
            depthColor[colorIndex + 2] = 255;
          }
          // 近すぎ 40cm-80cm(default mode)
          else if ( distance == depthStream.TooNearDepth ) {
            depthColor[colorIndex] = 0;
            depthColor[colorIndex + 1] = 255;
            depthColor[colorIndex + 2] = 0;
          }
          // 遠すぎ 3m(Near),4m(Default)-8m
          else if ( distance == depthStream.TooFarDepth ) {
            depthColor[colorIndex] = 255;
            depthColor[colorIndex + 1] = 0;
            depthColor[colorIndex + 2] = 0;
          }
          // 有効な距離データ
          else {
            depthColor[colorIndex] = 0;
            depthColor[colorIndex + 1] = 255;
            depthColor[colorIndex + 2] = 255;
          }
        }
      }

      return depthColor;
    }

    /// <summary>
    /// Windowが閉じられるときのイベント
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void Form1_FormClosing( object sender, FormClosingEventArgs e )
    {
      StopKinect( KinectSensor.KinectSensors[0] );
    }

    /// <summary>
    /// 距離カメラの通常/近接モード変更イベント
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    private void comboBoxRange_SelectedIndexChanged( object sender, EventArgs e )
    {
      try {
        KinectSensor.KinectSensors[0].DepthStream.Range = (DepthRange)comboBoxRange.SelectedIndex;
      }
      catch ( Exception ) {
        comboBoxRange.SelectedIndex = 0;
      }
    }
  }
}

File: VB/1_basic/WindowsForms/Form1.vb
Imports System
Imports System.Drawing
Imports System.Windows.Forms
Imports Microsoft.Kinect
Imports System.Drawing.Imaging
Imports System.Runtime.InteropServices

Partial Public Class Form1
    Inherits Form

    Private ReadOnly Bgr32BytesPerPixel As Integer = 4 ' ピクセルあたりのバイト数

    ''' <summary>
    ''' コンストラクタ
    ''' </summary>
    ''' <remarks></remarks>
    Public Sub New()
        InitializeComponent()

        Try
            ' Kinectが接続されているかどうかを確認する
            If (KinectSensor.KinectSensors.Count = 0) Then
                Throw New Exception("Kinectを接続してください")
            End If

            ' Kinectの動作を開始する
            Call StartKinect(KinectSensor.KinectSensors(0))
        Catch ex As Exception
            MessageBox.Show(ex.Message)
            Me.Close()
        End Try
    End Sub

    ''' <summary>
    ''' Kinectの動作を開始する
    ''' </summary>
    ''' <param name="kinect"></param>
    ''' <remarks></remarks>
    Private Sub StartKinect(ByVal kinect As KinectSensor)
        kinect.ColorStream.Enable()
        kinect.DepthStream.Enable()
        kinect.SkeletonStream.Enable()

        AddHandler kinect.AllFramesReady, AddressOf kinect_AllFramesReady

        kinect.Start()

        ' defaultモードとnearモードの切り替え
        Me.comboBoxRange.Items.Clear()
        For Each range In [Enum].GetValues(GetType(DepthRange))
            Me.comboBoxRange.Items.Add(range.ToString)
        Next
        Me.comboBoxRange.SelectedIndex = 0
    End Sub

    ''' <summary>
    ''' Kinectの動作を停止する
    ''' </summary>
    ''' <param name="kinect"></param>
    ''' <remarks></remarks>
    Private Sub StopKinect(ByVal kinect As KinectSensor)
        If kinect IsNot Nothing Then
            If kinect.IsRunning Then
                ' フレーム更新イベントを削除する
                RemoveHandler kinect.AllFramesReady, AddressOf kinect_AllFramesReady

                kinect.Stop()
                kinect.Dispose()
                kinect = Nothing

                Me.pictureBoxRgb.Image = Nothing
                Me.pictureBoxDepth.Image = Nothing
            End If
        End If
    End Sub

    ''' <summary>
    ''' RGBカメラ、距離カメラ、骨格のフレーム更新イベント
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub Kinect_AllFramesReady(sender As Object,
                                      e As Microsoft.Kinect.AllFramesReadyEventArgs)
        Try
            ' センサーのインスタンスを取得する
            Dim kinect As KinectSensor = CType(sender, KinectSensor)
            If kinect Is Nothing Then
                Exit Sub
            End If

            ' RGBカメラのフレームデータを取得する
            Using colorFrame As ColorImageFrame = e.OpenColorImageFrame
                If colorFrame IsNot Nothing Then
                    ' RGBカメラのピクセルデータを取得する
                    Dim colorPixel(colorFrame.PixelDataLength - 1) As Byte
                    colorFrame.CopyPixelDataTo(colorPixel)

                    ' ピクセルデータをビットマップに変換する
                    Dim _bitmap As New Bitmap(kinect.ColorStream.FrameWidth,
                                              kinect.ColorStream.FrameHeight,
                                              System.Drawing.Imaging.PixelFormat.Format32bppRgb)

                    Dim rect As New Rectangle(0, 0, _bitmap.Width, _bitmap.Height)
                    Dim data As BitmapData = _bitmap.LockBits(rect,
                                                              ImageLockMode.WriteOnly,
                                                              PixelFormat.Format32bppRgb)
                    Marshal.Copy(colorPixel, 0, data.Scan0, colorPixel.Length)
                    _bitmap.UnlockBits(data)

                    Me.pictureBoxRgb.Image = _bitmap
                End If
            End Using

            ' 書き込み用のビットマップデータを作成(32bit bitmap)
            ' 16bpp グレースケールは表示できない
            ' 距離カメラのフレームデータを取得する
            Using _depthFrame As DepthImageFrame = e.OpenDepthImageFrame
                If _depthFrame IsNot Nothing Then
                    Dim _bitmap As New Bitmap(kinect.DepthStream.FrameWidth,
                                              kinect.DepthStream.FrameHeight,
                                              System.Drawing.Imaging.PixelFormat.Format32bppRgb)

                    Dim rect As New Rectangle(0, 0, _bitmap.Width, _bitmap.Height)
                    Dim data As BitmapData = _bitmap.LockBits(rect, ImageLockMode.WriteOnly,
                                                            System.Drawing.Imaging.PixelFormat.Format32bppRgb)
                    Dim gray() As Byte = ConvertDepthColor(kinect, _depthFrame)
                    Marshal.Copy(gray, 0, data.Scan0, gray.Length)
                    _bitmap.UnlockBits(data)

                    Me.pictureBoxDepth.Image = _bitmap
                End If
            End Using

            ' スケルトンのフレームを取得する
            Using _skeletonFrame As SkeletonFrame = e.OpenSkeletonFrame
                If _skeletonFrame IsNot Nothing Then
                    Dim g As Graphics = Graphics.FromImage(Me.pictureBoxRgb.Image)

                    ' スケルトンのデータを取得する
                    Dim skeletons(_skeletonFrame.SkeletonArrayLength - 1) As Skeleton
                    _skeletonFrame.CopySkeletonDataTo(skeletons)

                    ' トラッキングされているスケルトンのジョイントを描画する
                    For Each _skeleton In skeletons
                        ' スケルトンがトラッキング状態(デフォルトモード)の場合は、ジョイントを描画する
                        If _skeleton.TrackingState = SkeletonTrackingState.Tracked Then
                            ' ジョイントを描画する
                            For Each _joint As Joint In _skeleton.Joints
                                ' ジョイントがトラッキングされていなければ次へ
                                If _joint.TrackingState <> JointTrackingState.Tracked Then
                                    Continue For
                                End If

                                ' スケルトンの座標を、RGBカメラの座標に変換して円を書く
                                Call DrawEllipse(kinect, g, _joint.Position)
                            Next
                            ' スケルトンが位置追跡(ニアモードの)の場合は、スケルトン位置(Center hip)を描画する
                        ElseIf _skeleton.TrackingState = SkeletonTrackingState.PositionOnly Then
                            ' スケルトンの座標を、RGBカメラの座標に変換して円を書く
                            Call DrawEllipse(kinect, g, _skeleton.Position)
                        End If
                    Next
                End If
            End Using
        Catch ex As Exception
            MessageBox.Show(ex.Message)
        End Try
    End Sub

    ''' <summary>
    ''' ジョイントの円を描く
    ''' </summary>
    ''' <param name="kinect"></param>
    ''' <param name="position"></param>
    ''' <remarks></remarks>
    Private Sub DrawEllipse(kinect As KinectSensor, g As Graphics, position As SkeletonPoint)
        Const R As Integer = 5

        ' スケルトンの座標を、RGBカメラの座標に変換する
        Dim _point As ColorImagePoint = kinect.MapSkeletonPointToColor(position, kinect.ColorStream.Format)

        g.DrawEllipse(New Pen(Brushes.Red, R),
                      New Rectangle(_point.X - R, _point.Y - R, R * 2, R * 2))
    End Sub

    ''' <summary>
    ''' 距離データをカラー画像に変換する
    ''' </summary>
    ''' <param name="kinect"></param>
    ''' <param name="depthFrame"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Private Function ConvertDepthColor(kinect As KinectSensor, depthFrame As DepthImageFrame) As Byte()
        Dim colorStream As ColorImageStream = kinect.ColorStream
        Dim depthStream As DepthImageStream = kinect.DepthStream

        ' 距離カメラのピクセルごとのデータを取得する
        Dim depthPixel(depthFrame.PixelDataLength - 1) As Short
        depthFrame.CopyPixelDataTo(depthPixel)

        ' 距離カメラの座標に対応するRGBカメラの座標を取得する(座標合わせ)
        Dim colorPoint(depthFrame.PixelDataLength - 1) As ColorImagePoint
        kinect.MapDepthFrameToColorFrame(depthStream.Format, depthPixel,
                                         colorStream.Format, colorPoint)

        Dim depthColor(depthFrame.PixelDataLength * Bgr32BytesPerPixel - 1) As Byte
        For index As Integer = 0 To depthPixel.Length - 1
            ' 距離カメラのデータから、プレイヤーIDと距離を取得する
            Dim player As Integer = depthPixel(index) And DepthImageFrame.PlayerIndexBitmask
            Dim distance As Integer = depthPixel(index) >> DepthImageFrame.PlayerIndexBitmaskWidth

            ' 変換した結果が、フレームサイズを超えることがあるため、小さいほうを使う
            Dim x As Integer = Math.Min(colorPoint(index).X, colorStream.FrameWidth - 1)
            Dim y As Integer = Math.Min(colorPoint(index).Y, colorStream.FrameHeight - 1)
            Dim colorIndex As Integer = ((y * depthFrame.Width) + x) * Bgr32BytesPerPixel

            If player <> 0 Then
                ' プレイヤーがいるピクセルの場合
                depthColor(colorIndex) = 255
                depthColor(colorIndex + 1) = 255
                depthColor(colorIndex + 2) = 255
            Else
                ' プレイヤーではないピクセルの場合
                If distance = depthStream.UnknownDepth Then
                    ' サポート外 0-40cm
                    depthColor(colorIndex) = 0
                    depthColor(colorIndex + 1) = 0
                    depthColor(colorIndex + 2) = 255
                ElseIf distance = depthStream.TooNearDepth Then
                    ' 近すぎ 40cm-80cm(default mode)
                    depthColor(colorIndex) = 0
                    depthColor(colorIndex + 1) = 255
                    depthColor(colorIndex + 2) = 0
                ElseIf distance = depthStream.TooFarDepth Then
                    ' 遠すぎ 3m(Near),4m(Default)-8m
                    depthColor(colorIndex) = 255
                    depthColor(colorIndex + 1) = 0
                    depthColor(colorIndex + 2) = 0
                Else
                    ' 有効な距離データ
                    depthColor(colorIndex) = 0
                    depthColor(colorIndex + 1) = 255
                    depthColor(colorIndex + 2) = 255
                End If
            End If
        Next

        Return depthColor
    End Function

    ''' <summary>
    ''' Windowsが閉じられるときのイベント
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub Form1_FormClosing(sender As Object,
                                  e As System.Windows.Forms.FormClosingEventArgs) Handles Me.FormClosing
        Call StopKinect(KinectSensor.KinectSensors(0))
    End Sub

    ''' <summary>
    ''' 距離カメラの通常/近接モード変更イベント
    ''' </summary>
    ''' <param name="sender"></param>
    ''' <param name="e"></param>
    ''' <remarks></remarks>
    Private Sub comboBoxRange_SelectionChanged(sender As System.Object,
                                               e As EventArgs)
        Try
            KinectSensor.KinectSensors(0).DepthStream.Range = CType(comboBoxRange.SelectedIndex, DepthRange)
        Catch ex As Exception
            comboBoxRange.SelectedIndex = 0
        End Try

    End Sub
End Class
